Addon-info: deep merge passed styles with default ones

jennvoss opened the pull request:
Issue: 
The styles option for addon-info was improperly documented
https://github.com/storybooks/storybook/issues/2144

## What I did
Updated addon-info docs to include the proper reference for the styles option and updated the example in cra-kitchen-sink

jennvoss committed:
Update styles option reference per https://github.com/storybooks/storybook/issues/2144

jennvoss committed:
Update addon-info styles example in cra-kitchen-sink

Hypnosphi reviewed (approved):
LGTM, thanks @jennvoss

Hypnosphi commented:
Please run `yarn test --core --update` to update snapshots of our example stories. You may need to bootstrap the project with `yarn bootstrap --core` if you didn't yet

jennvoss commented:
thanks @Hypnosphi, i will do that.  

Also found a deeper issue with this - if you don't include properties that contain nested objects from the stylesheet object (https://github.com/storybooks/storybook/blob/master/addons/info/src/components/Story.js#L19) such as "button", "header" , or "source" then the storybook will throw an error. Will try to look into this more over the weekend.

jennvoss committed:
Change styles prop to expect an object instead of a function

jennvoss committed:
Update cra-kitchen-sink example and snapshot

jennvoss committed:
Update styles reference in README

jennvoss commented:
Update: I changed the `styles` prop to accept an object instead of a function, and used nested-object-assign to only overwrite the properties that were passed. Hot reloading still works fine this way (https://github.com/storybooks/storybook/issues/2144 mentioned a concern about this). Updated the README, cra-kitchen-sink example and snapshots.

Hypnosphi commented on addons/info/src/components/Story.js:
```diff
@@ -382,7 +383,7 @@ Story.propTypes = {
   showHeader: PropTypes.bool,
   showSource: PropTypes.bool,
-  styles: PropTypes.func.isRequired,
+  styles: PropTypes.object, // eslint-disable-line react/forbid-prop-types
```
Let's rather make it `PropTypes.oneOfType(PropTypes.func, PropTypes.object)` and add a runtime check. If user supplies a function we call it, otherwise do a deep merge.

That way we won't introduce any breaking changes, all the old code will keep working

jennvoss replied:
good call, will update

Hypnosphi replied:
Oh, we actually do that runtime check already:
https://github.com/storybooks/storybook/blob/master/addons/info/src/index.js#L56

Let's just replace the default `s => s` function with deep merge

Hypnosphi commented:
I changed the base of this PR to `release/3.3` because now it introduces a new feature. If you need any help with conflicts resolution etc., feel free to contact us on slack: https://now-examples-slackin-nqnzoygycp.now.sh/

Hypnosphi commented:
Styles option as function is still worth documenting, and the example should indicate that default stylesheet is passed as argument. It can look like that:
```
styles: stylesheet => ({
  ...stylesheet,
  header: {
    ...stylesheet.header,
    h1: {
      ...stylesheet.header.h1,
      color: 'red'
    }
  }
})
```

This is more verbose than just passing an object that would be deep merged, but at the same time gives more control. This is kinda like Extend vs Full Control mode for custom webpack config: https://storybook.js.org/configurations/custom-webpack-config/

jennvoss committed:
Allow fn or object prop for styles for backward compatibility

danielduan reviewed (approved):
looks good, just add the style function backwards compatibility and we're good to go :)

jennvoss committed:
Pass stylesheet as the default arg to props.styles, update README and cra examples

jennvoss committed:
Merge branch 'release/3.3' into update-addon-info-docs

jennvoss committed:
Merge remote-tracking branch 'upstream/release/3.3' into update-addon-info-docs

Hypnosphi commented on addons/info/src/components/Story.js:
```diff
@@ -384,7 +384,7 @@ Story.propTypes = {
   showHeader: PropTypes.bool,
   showSource: PropTypes.bool,
-  styles: PropTypes.func.isRequired,
+  styles: PropTypes.oneOfType([PropTypes.func, PropTypes.object]),
```
Looks like this proptype may be left as is (we actually pass a function anyway)

jennvoss replied:
👍  done

Hypnosphi committed:
Merge branch 'release/3.3' into update-addon-info-docs

jennvoss committed:
Set styles to proptypes.func

Hypnosphi commented:
Please update the storyshots one more time

jennvoss commented:
When i run `yarn test --core --update`, everything passes and no updates are present. Little lost as to why it fails here. I'll make another change to the cra-kitchen-sink example and try again.

jennvoss committed:
Update snapshots

jennvoss commented:
Looks like i needed to run `yarn bootstrap --core` again as well
